components/nexus-repository/src/main/java/org/sonatype/nexus/repository/negativecache/NegativeCacheHandler.java

```diff
@@ -30,35 +36,33 @@
   @Nonnull
   @Override
   public Response handle(@Nonnull final Context context) throws Exception {
-    assert "GET".equals(context.getRequest().getAction());
-
-    NegativeCacheKey key = cacheKey(context);
-
-    final NegativeCacheFacet nfc = context.getRepository().facet(NegativeCacheFacet.class);
-
-    if (nfc.isNotFound(key)) {
-      return HttpResponses.notFound();
+    String action = context.getRequest().getAction();
+    if (!HttpMethods.GET.equals(action)) {
+      return context.proceed();
     }
+    NegativeCacheFacet negativeCache = context.getRepository().facet(NegativeCacheFacet.class);
+    NegativeCacheKey key = negativeCache.getCacheKey(context);
 
-    final Response response = context.proceed();
-
-    if (indicatesNotFound(response)) {
-      nfc.cacheNotFound(key);
-      return HttpResponses.notFound();
+    Response response;
+    Status status = negativeCache.get(key);
+    if (status == null) {
+      response = context.proceed();
+      if (isNotFound(response)) {
+        negativeCache.put(key, response.getStatus());
+      }
+      else if (response.getStatus().isSuccessful()) {
+        negativeCache.invalidate(key);
```

Comment:
do we need to invalidate in this case? we could only get here if this wasn't in the negativeCache in the first place so it looks like this would always be a no-op anyhow